### PR TITLE
[KRYS-5664] Add platform filter in earn list v2

### PR DESCRIPTION
**Pull request by MrQKyber**

### CODE REVIEW

**Comment by tungnguyen20** [comment]
- Missing earn type filter (Stake/Lend) 
- Not multichoice options

File: EarnModule/EarnModule/Resources/Assets.xcassets/all_platform_icon.imageset/Contents.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      "filename" : "all@3x.png",
+      "idiom" : "universal",
+      "scale" : "3x"
+    }
+  ],
+  "info" : {
+    "author" : "xcode",
+    "version" : 1
+  }
+}
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Resources/Assets.xcassets/all_platform_icon.imageset/Contents.json:23]: Why the icon is not centerized?

File: EarnModule/EarnModule/Resources/Localizable.strings [hunk 2]
```diff
@@ -62,6 +62,7 @@
 "your.staking.balance.is.not.suffcient" = "Your staking balance is not sufficient";
 "connect.wallet" = "Connect Wallet";
 "insufficient.quote.balance" = "Insufficient %@ balance";
+"All Networks" = "All Networks";
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Resources/Localizable.strings:65]: I think it should be "All platforms"

File: EarnModule/EarnModule/Resources/Strings.swift [hunk 3]
```diff
@@ -109,6 +109,7 @@ struct Strings {
     static let youWillReceiveYourIn = "you.will.receive.your.in".toBeLocalised()
     static let amountQuoteTokenUsedForFee = "use.amount.of.quote.token.for.fee.message".toBeLocalised()
     static let yourStakingBalanceIsNotSufficient = "your.staking.balance.is.not.suffcient".toBeLocalised()
+    static let allNetworks = "All Networks".toBeLocalised()
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Resources/Strings.swift:112]: Please follow the convention.

File: EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import UIKit
+import Services
+
+protocol PlatformFilterViewControllerDelegate: class {
+    func didSelectPlatform(viewController: PlatformFilterViewController, selected: EarnPlatform?)
+}
+
+class PlatformFilterViewModel {
+    var dataSource: [EarnPlatform]
+    var selected: EarnPlatform?
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift:17]: When the list is multichoice, selected will be an array

File: EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        self.selected = selected
+    }
+    
+    func platformForRow(row: Int) -> EarnPlatform? {
+        if row == 0 {
+            return nil
+        }
+        let index = row - 1
+        return dataSource[index]
+    }
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift:30]: So this func should return an array

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 6]
```diff
@@ -134,7 +170,25 @@ class EarnListViewController: InAppBrowsingViewController {
...
+        viewController.viewModel = viewModel
+        viewController.delegate = self
+        let sheetOptions = SheetOptions(pullBarHeight: 0)
+        let sheet = SheetViewController(controller: viewController, sizes: [.intrinsic], options: sheetOptions)
+        present(sheet, animated: true)
+    }
+    
+    private func updateUIPlatformFilterButton() {
+        guard let selectedPlatform = selectedPlatform else {
+            platformFilterButton.setTitle("All Networks", for: .normal)
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:187]: Localize

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 7]
```diff
@@ -62,8 +65,27 @@ class EarnListViewController: InAppBrowsingViewController {
...
+            self.emptyIcon.image = UIImage(named: "empty-search-token")
+            self.emptyLabel.text = Strings.noRecordFound
+          }
+      }
+      
+      if let selectedPlatform = selectedPlatform {
+          self.displayDataSource = self.displayDataSource.filter { element in
+              return element.earnPoolModel.platforms.contains(selectedPlatform)
+          }
+      }
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:85]: I think this logic should be in ViewModel

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 8]
```diff
@@ -100,6 +122,20 @@ class EarnListViewController: InAppBrowsingViewController {
...
+    private func getAllPlatform() -> [EarnPlatform] {
+        var platformSet = Set<EarnPlatform>()
+        
+        dataSource.forEach { item in
+            item.earnPoolModel.platforms.forEach { element in
+                platformSet.insert(element)
+            }
+        }
+        return Array(platformSet).sorted { (left, right) -> Bool in
+            return left.name > right.name
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:135]: Must be `<` to sort A-Z

File: EarnModule/EarnModule/Resources/Assets.xcassets/all_platform_icon.imageset/Contents.json (hunk 1 above)
Comment (MrQKyber) [reply EarnModule/EarnModule/Resources/Assets.xcassets/all_platform_icon.imageset/Contents.json:23]: plz check the design file

File: EarnModule/EarnModule/Resources/Localizable.strings (hunk 2 above)
Comment (MrQKyber) [reply EarnModule/EarnModule/Resources/Localizable.strings:65]: Plz check the design file

File: EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift (hunk 4 above)
Comment (MrQKyber) [reply EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift:17]: this is single choice

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift (hunk 7 above)
Comment (MrQKyber) [reply EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:85]: Plz ask Thang for this issue, this class has no view model

File: EarnModule/EarnModule/Resources/Assets.xcassets/all_platform_icon.imageset/Contents.json (hunk 1 above)
Comment (tungnguyen20) [reply EarnModule/EarnModule/Resources/Assets.xcassets/all_platform_icon.imageset/Contents.json:23]: I think when seeing some thing wrong, as the PIC of the story, you should raise the issue with design team.

File: EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift (hunk 4 above)
Comment (tungnguyen20) [reply EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift:17]: Have you checked the design?

Comment (tungnguyen20) [reply EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift:17]: And please check this line in the notion page
> User can multi-select earning platform

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 9]
```diff
@@ -12,253 +12,326 @@ import Dependencies
...
+        selectedPlatform = Set(getAllPlatform())
+        fetchData(chainId: currentSelectedChain == .all ? nil : currentSelectedChain.getChainId())
+        Timer.scheduledTimer(withTimeInterval: 15.0, repeats: true) { [weak self] _ in
+            self?.fetchData(chainId: self?.currentSelectedChain == .all ? nil : self?.currentSelectedChain.getChainId(), isAutoReload: true)
+        }
+        setupUI()
+    }
+    
+    
+
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:50]: Please remove redundant empty lines

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 10]
```diff
@@ -12,253 +12,326 @@ import Dependencies
...
+    weak var delegate: EarnListViewControllerDelegate?
+    
+    @IBOutlet weak var platformFilterButton: UIButton!
+    @IBOutlet weak var emptyIcon: UIImageView!
+    @IBOutlet weak var emptyLabel: UILabel!
+    var dataSource: [EarnPoolViewCellViewModel] = []
+    var displayDataSource: [EarnPoolViewCellViewModel] = []
+    var timer: Timer?
+    var currentSelectedChain: ChainType = AppState.shared.isSelectedAllChain ? .all : AppState.shared.currentChain
+    var selectedPlatform: Set<EarnPlatform>!
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:37]: Should be `selectedPlatforms'

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 11]
```diff
@@ -12,253 +12,326 @@ import Dependencies
...
+        currentSelectedChain = .all
+        fetchData()
+    }
+    
+    func setupUI() {
+        self.searchTextField.setPlaceholder(text: Strings.searchToken, color: AppTheme.current.secondaryTextColor)
+        self.tableView.registerCellNib(EarnPoolViewCell.self)
+    }
+    
+    private func isSelectedAllPlatform() -> Bool {
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:70]: `isSelectedAllPlatforms`

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 12]
```diff
@@ -12,253 +12,326 @@ import Dependencies
...
+    
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        timer?.invalidate()
+        timer = Timer.scheduledTimer(timeInterval: 0.5, target: self, selector: #selector(doSearch), userInfo: nil, repeats: false)
+        return true
+    }
+    
+    func filterDataSource(text: String) {
+        self.displayDataSource = self.dataSource.filter({ viewModel in
+            let containSymbol = viewModel.earnPoolModel.token.symbol.lowercased().contains(text.lowercased())
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:303]: Please trim search text

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 13]
```diff
@@ -12,253 +12,326 @@ import Dependencies
...
+        if !isAutoReload {
+            self.displayDataSource.forEach { viewModel in
+                viewModel.isExpanse = false
+            }
+            reloadUI()
+        }
+        let service = EarnServices()
+        if !isAutoReload {
+            showLoading()
+        }
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:109]: Why not move `showLoading` to the above `if !isAutoReload {`

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift [hunk 14]
```diff
@@ -12,253 +12,326 @@ import Dependencies
...
+        viewController.viewModel = viewModel
+        viewController.delegate = self
+        let sheetOptions = SheetOptions(pullBarHeight: 0)
+        let sheet = SheetViewController(controller: viewController, sizes: [.intrinsic], options: sheetOptions)
+        present(sheet, animated: true)
+    }
+    
+    private func updateUIPlatformFilterButton() {
+        if isSelectedAllPlatform() {
+            platformFilterButton.setTitle(Strings.allNetworks, for: .normal)
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:208]: `all platforms`

File: EarnModule/EarnModule/Cells/PlatformCell.swift [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        // Initialization code
+    }
+    
+    func updateCell(platform: EarnPlatform?, isSelected: Bool) {
+        guard let platform = platform else {
+            iconImageView.image = Images.allNetworkIcon
+            nameLabel.text = Strings.allNetworks
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Cells/PlatformCell.swift:25]: `Strings.allPlatforms`

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift (hunk 13 above)
Comment (MrQKyber) [reply EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:109]: Ask Thang for this logic

File: EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift (hunk 5 above)
Comment (MrQKyber) [reply EarnModule/EarnModule/ViewControllers/PlatformFilterViewController.swift:30]: Can not understand this getter function is used for getting one item

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift (hunk 12 above)
Comment (MrQKyber) [reply EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:303]: what do you mean ? text has been lowercased ??

Comment (tungnguyen20) [reply EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:303]: I mean sometimes user type with spaces around the text

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift (hunk 7 above)
Comment (tungnguyen20) [reply EarnModule/EarnModule/ViewControllers/EarnListViewController.swift]: Let's help Thang improve it :)

File: EarnModule/EarnModule/ViewControllers/EarnListViewController.swift (hunk 13 above)
Comment (tungnguyen20) [reply EarnModule/EarnModule/ViewControllers/EarnListViewController.swift:109]: Let's help Thang improve it :)

**Review by tungnguyen20 — APPROVED** [review]

**Pull request merged**